### PR TITLE
Refactor Object Detection inference to use new Model Trainer type

### PR DESCRIPTION
**Pull request by nickjong**

### CODE REVIEW

File: src/toolkits/object_detection/od_darknet_yolo_model_trainer.cpp [hunk 1]
```diff
@@ -252,6 +260,7 @@ EncodedInputBatch EncodeDarknetYOLO(InputBatch input_batch,
   result.iteration_id = input_batch.iteration_id;
   result.images = std::move(input_batch.images);
   result.annotations = std::move(input_batch.annotations);
+  result.image_sizes = std::move(input_batch.image_sizes);
```
Comment (shreyajain17) [inline src/toolkits/object_detection/od_darknet_yolo_model_trainer.cpp]: What is this for?

File: src/toolkits/object_detection/od_darknet_yolo_model_trainer.cpp [hunk 2]
```diff
@@ -388,7 +430,82 @@ DarknetYOLOModelTrainer::DarknetYOLOModelTrainer(
...
+    std::unique_ptr<data_iterator> test_data, size_t batch_size,
+    float confidence_threshold, float iou_threshold) {
+  // Wrap the data_iterator to incorporate into a Combine pipeline.
+  auto iterator = std::make_shared<DataIterator>(std::move(test_data),
+                                                 batch_size, /* offset */ 0);
+
+  // No labels to encode. Just pass the annotations through for potential
+  // evaluation.
+  auto trivial_encoder = [](InputBatch input_batch) {
+    EncodedInputBatch result;
```
Comment (shreyajain17) [inline src/toolkits/object_detection/od_darknet_yolo_model_trainer.cpp]: Why are we doing this?

File: src/toolkits/object_detection/od_darknet_yolo_model_trainer.cpp (hunk 1 above)
Comment (nickjong) [reply src/toolkits/object_detection/od_darknet_yolo_model_trainer.cpp]: Good question. For prediction, we want to yield predicted bounding boxes in image coordinates. Without this, we don't have a convenient way to scale our (normalized) bounding box predictions.

Comment (shreyajain17) [reply src/toolkits/object_detection/od_darknet_yolo_model_trainer.cpp]: Makes sense

File: src/toolkits/object_detection/od_darknet_yolo_model_trainer.cpp (hunk 2 above)
Comment (nickjong) [reply src/toolkits/object_detection/od_darknet_yolo_model_trainer.cpp]: The augmenter currently outputs an `InputBatch` but the model backend wants an `EncodedInputBatch`, which is basically `InputBatch` plus some stuff for loss labels that don't apply to the inference case. So we just need to copy all the contents of the `InputBatch` into the `EncodedInputBatch`, since there are no loss labels to encode.
